app/routes/user/endorsements.server.ts

```diff
@@ -0,0 +1,265 @@
+    if (!user)
+      throw new Response('Forbidden', {
+        status: 403,
+      })
+    return new this(user.sub, user.email, user.groups)
+  }
+
+  /**
+   * Request a pending endorsement.
+   *
+   * Throws an HTTP error if:
+   *   - The requested endorser is the same as the current user
+   *   - User `endorserSub` does not exist
+   *   - User `endorserSub` is not in the GCN Circulars submitters group
+   *   - User `endorserSub` does not have an email attribute
+   *   - There is an existing endorsement with the same requester and submitter sub
+   */
+  async createEndorsementRequest(endorserSub: string) {
+    if (endorserSub === this.#sub)
+      throw new Response(
+        'Users cannot request themselves as the endorser of their own request',
+        {
+          status: 400,
+        }
+      )
+
+    const escapedEndorserString = endorserSub.replaceAll('"', '\\"')
+    const user = (
+      await cognitoIdentityProviderClient.send(
+        new ListUsersCommand({
+          UserPoolId: process.env.COGNITO_USER_POOL_ID,
+          Filter: `sub = "${escapedEndorserString}"`,
+        })
+      )
+    )?.Users?.[0]
+    if (!user)
+      throw new Response('Requested user does not exist', {
+        status: 400,
+      })
+
+    const getGroupsCommand = new AdminListGroupsForUserCommand({
+      Username: user.Username,
+      UserPoolId: process.env.COGNITO_USER_POOL_ID,
+    })
+    const groupsResponse = await cognitoIdentityProviderClient.send(
+      getGroupsCommand
+    )
+
+    if (
+      !groupsResponse.Groups?.find(
+        ({ GroupName }) => GroupName === 'gcn.nasa.gov/circular-submitter'
+      )
+    )
+      throw new Response('User is not in the submitters group', {
+        status: 400,
+      })
+
+    const endorserEmail = user.Attributes?.find((x) => x.Name == 'email')?.Value
+    if (!endorserEmail)
+      throw new Response('Requested user does not have an email', {
+        status: 400,
+      })
+
+    const db = await tables()
+
+    await db.circular_endorsements.update({
+      Key: {
+        requestorSub: this.#sub,
+        endorserSub: escapedEndorserString,
+      },
+      UpdateExpression:
+        'set #status = :status, #endorserEmail = :endorserEmail, #requestorEmail = :requestorEmail, #created = :created',
+      ExpressionAttributeNames: {
+        '#status': 'status',
+        '#endorserEmail': 'endorserEmail',
+        '#requestorEmail': 'requestorEmail',
+        '#created': 'created',
+        '#requestorSub': 'requestorSub',
+        '#endorserSub': 'endorserSub',
+      },
+      ExpressionAttributeValues: {
+        ':status': 'pending',
+        ':endorserEmail': endorserEmail,
+        ':requestorEmail': this.#currentUserEmail,
+        ':created': Date.now(),
+        ':endorserSub': escapedEndorserString,
+        ':requestorSub': this.#sub,
+      },
+      ConditionExpression:
+        'NOT (#endorserSub = :endorserSub and #requestorSub = :requestorSub)',
```

Comment:
What? No. This should be checking that if the endorsement request exists, then its status is "pending".

Reply:
Why are we checking pending?  This is correct, since it is true if there isnt an existing one with a matching endorser and requestor sub.

Reply:
So that creating an endorsement is idempotent.